File: spec/check_spec_anchors.py
#!/usr/bin/env python3
"""Verify conformance references and relative Markdown links."""

from __future__ import annotations

from collections import defaultdict
from pathlib import Path
import re
import sys
from urllib.parse import unquote


HEADING = re.compile(r"^#{1,6}\s+(.+?)\s*#*\s*$", re.MULTILINE)
LINK = re.compile(r"(?<!!)\[[^]]*\]\(([^)]+)\)")


def github_slug(text: str) -> str:
    """Implement the GitHub heading-slug rules needed by this repository."""

    text = re.sub(r"<[^>]+>", "", text)
    text = re.sub(r"[`*_~]", "", text.strip().lower())
    text = re.sub(r"[^\w\- ]", "", text, flags=re.UNICODE)
    return text.replace(" ", "-")


def anchors(path: Path) -> set[str]:
    counts: dict[str, int] = defaultdict(int)
    result: set[str] = set()
    for match in HEADING.finditer(path.read_text(encoding="utf-8")):
        base = github_slug(match.group(1))
        index = counts[base]
        counts[base] += 1
        result.add(base if index == 0 else f"{base}-{index}")
    return result


def prose_without_fences(text: str) -> str:
    return re.sub(r"^```.*?^```\s*$", "", text, flags=re.MULTILINE | re.DOTALL)


def link_target(raw: str) -> str:
    raw = raw.strip()
    if raw.startswith("<") and ">" in raw:
        return raw[1 : raw.index(">")]
    return raw.split(maxsplit=1)[0]


def check_reference(
    source: Path,
    destination: str,
    cache: dict[Path, set[str]],
    failures: list[str],
) -> None:
    if not destination or destination.startswith(("http://", "https://", "mailto:")):
        return
    target_text, separator, fragment = destination.partition("#")
    target_text = unquote(target_text).split("?", 1)[0]
    if target_text and Path(target_text).suffix not in {".md", ".hd", ".tsv"}:
        # The grammar and type prose contains code-like `[T](value)` forms.
        # They are not Markdown file links; repository links use an explicit
        # extension or a same-document fragment.
        return
    target = source if not target_text else (source.parent / target_text).resolve()
    if not target.exists() or not target.is_file():
        failures.append(f"{source}: missing link target {destination}")
        return
    if separator and fragment:
        decoded = unquote(fragment)
        known = cache.setdefault(target, anchors(target))
        if decoded not in known:
            failures.append(f"{source}: missing link anchor {destination}")


def main() -> int:
    if len(sys.argv) != 3:
        print("usage: check_spec_anchors.py SPEC_DIR CASES_TSV", file=sys.stderr)
        return 2
    spec_dir = Path(sys.argv[1]).resolve()
    repo_dir = spec_dir.parent
    failures: list[str] = []
    cache: dict[Path, set[str]] = {}

    for row in Path(sys.argv[2]).read_text(encoding="utf-8").splitlines()[1:]:
        fixture, _, _, reference = row.split("\t")
        file_text, separator, fragment = reference.partition("#")
        target = spec_dir / file_text
        if not target.is_file():
            failures.append(f"{fixture}: missing specification {reference}")
        elif separator and fragment not in cache.setdefault(target, anchors(target)):
            failures.append(f"{fixture}: missing specification anchor {reference}")

    for directory in (spec_dir, repo_dir / "guide", repo_dir / "future-work"):
        for source in sorted(directory.rglob("*.md")):
            prose = prose_without_fences(source.read_text(encoding="utf-8"))
            for match in LINK.finditer(prose):
                check_reference(source, link_target(match.group(1)), cache, failures)

    if failures:
        print("unresolved specification links:", file=sys.stderr)
        for failure in failures:
            print(f"- {failure}", file=sys.stderr)
        return 1
    print("specification links and anchors passed")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: spec/check-spec-anchors.ts
import { readFile, readdir, stat } from "node:fs/promises";
import { dirname, extname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const headingPattern = /^#{1,6}\s+(.+?)\s*#*\s*$/gm;
const linkPattern = /(?<!!)\[[^\]]*\]\(([^)]+)\)/g;

function githubSlug(text: string): string {
  return text
    .replaceAll(/<[^>]+>/g, "")
    .trim()
    .toLowerCase()
    .replaceAll(/[`*_~]/g, "")
    .replaceAll(/[^\p{L}\p{N}_\- ]/gu, "")
    .replaceAll(" ", "-");
}

async function anchors(path: string): Promise<Set<string>> {
  const counts = new Map<string, number>();
  const result = new Set<string>();
  const text = await readFile(path, "utf8");
  for (const match of text.matchAll(headingPattern)) {
    const base = githubSlug(match[1]!);
    const index = counts.get(base) ?? 0;
    counts.set(base, index + 1);
    result.add(index === 0 ? base : `${base}-${index}`);
  }
  return result;
}

function proseWithoutFences(text: string): string {
  return text.replaceAll(/^```.*?^```\s*$/gms, "");
}

function linkTarget(raw: string): string {
  const value = raw.trim();
  if (value.startsWith("<") && value.includes(">")) return value.slice(1, value.indexOf(">"));
  return value.split(/\s+/, 1)[0]!;
}

async function isFile(path: string): Promise<boolean> {
  try {
    return (await stat(path)).isFile();
  } catch {
    return false;
  }
}

async function markdownPaths(directory: string): Promise<string[]> {
  const paths: string[] = [];
  for (const entry of await readdir(directory, { withFileTypes: true })) {
    const path = resolve(directory, entry.name);
    if (entry.isDirectory()) paths.push(...(await markdownPaths(path)));
    else if (entry.isFile() && entry.name.endsWith(".md")) paths.push(path);
  }
  return paths.sort();
}

async function knownAnchors(path: string, cache: Map<string, Set<string>>): Promise<Set<string>> {
  const found = cache.get(path);
  if (found) return found;
  const loaded = await anchors(path);
  cache.set(path, loaded);
  return loaded;
}

async function checkReference(
  source: string,
  destination: string,
  cache: Map<string, Set<string>>,
): Promise<string | undefined> {
  if (!destination || /^(?:https?:\/\/|mailto:)/.test(destination)) return undefined;
  const [rawTarget = "", rawFragment] = destination.split("#", 2);
  const targetText = decodeURIComponent(rawTarget).split("?", 1)[0]!;
  if (targetText && ![".md", ".hd", ".tsv"].includes(extname(targetText))) return undefined;
  const target = targetText ? resolve(dirname(source), targetText) : source;
  if (!(await isFile(target))) return `${source}: missing link target ${destination}`;
  if (rawFragment) {
    const fragment = decodeURIComponent(rawFragment);
    if (!(await knownAnchors(target, cache)).has(fragment))
      return `${source}: missing link anchor ${destination}`;
  }
  return undefined;
}

async function main(args: readonly string[]): Promise<number> {
  if (args.length !== 2) {
    console.error("usage: check-spec-anchors.ts SPEC_DIR CASES_TSV");
    return 2;
  }
  const [specDirectory, casesPath] = args.map((path) => resolve(path));
  const repository = resolve(specDirectory!, "..");
  const failures: string[] = [];
  const cache = new Map<string, Set<string>>();
  const rows = (await readFile(casesPath!, "utf8")).split(/\r?\n/).slice(1).filter(Boolean);
  for (const row of rows) {
    const [fixture, , , reference = ""] = row.split("\t");
    const [fileText = "", fragment] = reference.split("#", 2);
    const target = resolve(specDirectory!, fileText);
    if (!(await isFile(target))) failures.push(`${fixture}: missing specification ${reference}`);
    else if (fragment && !(await knownAnchors(target, cache)).has(fragment))
      failures.push(`${fixture}: missing specification anchor ${reference}`);
  }
  const directories = [
    specDirectory!,
    resolve(repository, "guide"),
    resolve(repository, "future-work"),
  ];
  const sources = (await Promise.all(directories.map(markdownPaths))).flat();
  for (const source of sources) {
    const prose = proseWithoutFences(await readFile(source, "utf8"));
    for (const match of prose.matchAll(linkPattern)) {
      const failure = await checkReference(source, linkTarget(match[1]!), cache);
      if (failure) failures.push(failure);
    }
  }
  if (failures.length > 0) {
    console.error(
      `unresolved specification links:\n${failures.map((failure) => `- ${failure}`).join("\n")}`,
    );
    return 1;
  }
  console.log("specification links and anchors passed");
  return 0;
}

const invokedPath = process.argv[1] ? resolve(process.argv[1]) : "";
if (invokedPath === fileURLToPath(import.meta.url))
  process.exitCode = await main(process.argv.slice(2));
